src/librustdoc/clean/types.rs

```diff
@@ -1460,7 +1460,9 @@ crate enum Type {
     /// Structs/enums/traits (most that would be an `hir::TyKind::Path`).
     ResolvedPath {
         path: Path,
-        param_names: Option<Vec<GenericBound>>,
+        /// If `param_names` is `Some`, this path is a trait object and the Vecs repsresent
+        /// `(generic bounds, generic parameters)`
+        param_names: Option<(Vec<GenericBound>, Vec<GenericParamDef>)>,
```

Comment:
I somehow expected to get questions on the code because It's *really* weird, I know :sweat_smile: 

> Also, it seems odd to me that data specific to dyn Traits is stored as part of ResolvedPath. I noticed that you said in #84814 (comment) that you were planning to add a new Type variant but decided to instead change the existing ResolvedPath field. I think your original idea would actually be a better approach, and would make the code easier to understand.

Yes I agree that this is definitely cleaner and easier to understand. I will change that!

> I feel a bit confused about this. PathSegments (Path is mostly a Vec<PathSegment>) have generic arguments in them, so why are generic arguments both there and here?

This is the "hack" that rustdoc currently uses. It uses the `Path` to display trait objects. Before this PR, the `param_names` field would indicate if this `Path` is a trait object. If the `param_names` is `Some`, then this `Path` is a trait object and it's bounds are *this* path, and all the bounds in the `param_names` (`param_names` is basically a list of [these](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.PolyTraitRef.html#structfield.trait_ref) elements, where the first trait ref is special-cased and becomes *this* path) field. The second list in the tuple, was a list containing [these](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.PolyTraitRef.html#structfield.bound_generic_params) elements, so it includes all lifetimes that were provided using `for<...>`.

Reply:
> Could you explain a bit why you added the new field?

The `bounds` part does not include the higher ranked lifetimes. If you look at the doc for the rustic [WhereBoundPredicate](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.WhereBoundPredicate.html), you'll see that there's a `bounds` field, which directly translates to the `bounds` field in the cleaned `WherePredicate`, and a `bound_generic_params`, which includes all lifetimes provided by `for<>` (as written in the documentation). The latter field was previously discarded while cleaning the predicate, so I added a field to keep the information about these lifetimes.